emote/algos/genrl/proxies.py

```diff
@@ -0,0 +1,46 @@
+from typing import Dict, Optional
+
+import numpy as np
+import torch
+import torch.nn as nn
+
+from emote.memory.memory import TableMemoryProxy
+from emote.memory.table import Table
+from emote.typing import AgentId, DictObservation, DictResponse
+
+
+class MemoryProxyWithEncoder(TableMemoryProxy):
+    def __init__(
+        self,
+        table: Table,
+        encoder: nn.Module,
+        minimum_length_threshold: Optional[int] = None,
+        use_terminal: bool = False,
+        input_key: str = "obs",
+    ):
+        super().__init__(table, minimum_length_threshold, use_terminal)
+        self.encoder = encoder
+        self._input_key = input_key
+
+    def add(
+        self,
+        observations: Dict[AgentId, DictObservation],
+        responses: Dict[AgentId, DictResponse],
+    ):
+        updated_responses = {}
+        for agent_id, response in responses.items():
+            actions = np.array(response.list_data["actions"])
+            if np.size(actions) == 0:
+                updated_responses.update({agent_id: response})
+            else:
+                actions = torch.from_numpy(actions).unsqueeze(dim=0).to(torch.float)
+                obs = torch.from_numpy(
+                    observations[agent_id].array_data[self._input_key]
+                )
+                obs = obs.unsqueeze(dim=0).to(torch.float)
```

Comment:
All these squeeze/unsqueeze feel a bit ugly. I wonder if we can make it a bit nicer somehow.

Reply:
They are all fixed